coordinator/src/main/scala/filodb.coordinator/queryengine2/HighAvailabilityPlanner.scala

```diff
@@ -0,0 +1,97 @@
+package filodb.coordinator.queryengine2
+
+import com.typesafe.config.{Config, ConfigFactory}
+import com.typesafe.scalalogging.StrictLogging
+
+import filodb.core.{DatasetRef, SpreadProvider}
+import filodb.query.{LogicalPlan, PromQlInvocationParams, PromQlQueryParams, QueryContext}
+import filodb.query.exec.{ExecPlan, InProcessPlanDispatcher, PromQlExec, StitchRvsExec}
+
+/**
+  * HighAvailabilityPlanner responsible for using underlying local planner and FailureProvider
+  * to come up with a plan that orchestrates query execution between multiple
+  * replica clusters. If there are failures in one cluster then query is routed
+  * to other cluster.
+  *
+  * @param dsRef dataset
+  * @param spreadProvider used to get spread
+  * @param localPlanner the planner to generate plans for local pod
+  * @param failureProvider the provider that helps route plan execution to HA cluster
+  * @param queryEngineConfig config that determines query engine behavior
+  */
+class HighAvailabilityPlanner(dsRef: DatasetRef,
+                              localPlanner: QueryPlanner,
+                              failureProvider: FailureProvider,
+                              spreadProvider: SpreadProvider,
+                              queryEngineConfig: Config = ConfigFactory.empty) extends QueryPlanner with StrictLogging {
+
+  import QueryFailureRoutingStrategy._
+
+  /**
+    * Converts Route objects returned by FailureProvider to ExecPlan
+    */
+  private def routeExecPlanMapper(routes: Seq[Route], rootLogicalPlan: LogicalPlan,
+                                  options: QueryContext, lookBackTime: Long): ExecPlan = {
+
+    val execPlans: Seq[ExecPlan] = routes.map { route =>
+      route match {
+        case route: LocalRoute => if (route.timeRange.isEmpty)
+          localPlanner.materialize(rootLogicalPlan, options)
+        else
+          localPlanner.materialize(
+            copyWithUpdatedTimeRange(rootLogicalPlan, route.asInstanceOf[LocalRoute].timeRange.get, lookBackTime),
+            options)
+        case route: RemoteRoute =>
+          val timeRange = route.timeRange.get
+          val queryParams = options.origQueryParams.asInstanceOf[PromQlQueryParams]
+          val routingConfig = queryEngineConfig.getConfig("routing")
+          val promQlInvocationParams = PromQlInvocationParams(routingConfig, queryParams.promQl,
+            timeRange.startInMillis / 1000, queryParams.step, timeRange.endInMillis / 1000,
```

Comment:
This division by 1000 really confuses me. Some comments will help.

Reply:
PromQL params are in seconds. Added comment.